### ISSUE TITLE
Use libraries that support task migration

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#7780]
Among the approaches that have been proposed to address the fact that future supercomputers will consist of very very large numbers of cores is the idea that one should decompose programs into many small tasks that can be migrated between cores on different nodes for good load balancing. In other words, one would create very large DAGs of tasks for which a scheduler can find a suitable resource on the current node or elsewhere.

In some sense, we already do this in a good number of places by using `parallel_for` and `WorkStream`. But both of these only allow running on the local machine because the functions being called are associated with data stored in local memory. To make tasks migratable, one needs to adopt a "functional programming" style where everything a function needs is passed in through the input arguments (generally by-value), and everything the function does is returned through the return variable. The function cannot depend on other objects located somewhere in memory. The function also can't modify anything else. If functions satisfy this style, they can be run anywhere (on any available resource) as long as the input and output are serializable because then they can be migrated.

This is impracticable. But a task that reads and writes from other memory can be split into three:
* one that packs up all data necessary into a structure and that needs to run locally; 
* one that receives the packed data as argument, and that can run anywhere as long as the structure of data is serializable; the returned object can then also be migrated;
* one that unpacks the returned object and puts it where necessary; again, this task needs to run locally.

It would be interesting to think about what it would take to implement such a workflow for some parts of the library. I suspect that some of things we do with `WorkStream` could fit this model, for example for assembly. In follow-up comments, I may link to some libraries that facilitate this scheduling and data migration and that could form the basis for this kind of work.

**Comment by masterleinad** [comment]
There is https://github.com/cpp-taskflow/cpp-taskflow we might want to look into.

**Comment by kronbichler** [comment]
Overall I like this idea and agree that this would be an interesting addition for some of our applications. Yet, if there is something about shared-memory parallelism that I learned during the last years, then that DAG-based tasking is really meant for decisions that need to be made dynamically. Whenever there is something that is essentially static, like a for loop with fixed amount of work per item, or an item that can be load-balanced a priori to yield sub-tasks that are of equal size and that can be run independently, then this simple static variant is better than trying to do it dynamically. Part of the reason is that static decisions tend to favor contiguous memory access (that can be easily prefetched), and that static decisions made in the same way everywhere are the only way to optimize for NUMA.

Thus, what I would favor doing is to reserve dynamically scheduled tasks (that might even migrate over to another MPI rank) for things that are inherently non-static, and provide a static solution for the basic stuff. The more cores are added, the more difficult it becomes to fill them all with a dynamic solution vs a static split of for loops. So, to summarize my point, I definitely like this idea, but it will be a complement to a high-performance static solution that we need to provide as well.

**Comment by bangerth** [comment]
I agree with everything @kronbichler writes (as I usually do :-) ). For the moment, I'm mostly interested in experiments to see whether any of this is feasible at all.

I'll note that the functional programming approach would also make it substantially simpler to run things on GPUs.

**Comment by bangerth** [comment]
Here is one option, Legion: https://legion.stanford.edu/

I've looked through parts of the tutorial. I know that it supports migrating tasks across memory spaces/MPI nodes, but haven't seen a tutorial program. We'd have to look deeper.
